### PR TITLE
Consolidate descriptions linter functions into one

### PR DESCRIPTION
**Pull request by queengooborg**
Now that the descriptions linter has been merged (yay!), I'm opening up this PR to consolidate the code a little bit.  Essentially, this PR consolidates the four functions testing for each type of description into a single function, which calls out to another function that performs both the check and the logger output based upon a set of arguments.  This is to help keep the number of code lines to a minimum, and reduce the number of iterations through data the linter must perform.

I'm open to suggestions or revisions!

### CODE REVIEW

**Review by alshamiri5 — APPROVED** [review]

**Comment by ddbeck** [comment]
Hi @queengooborg, Florian and I talked about this in our most recent 1:1 (we're trying to get back in the habit of unblocking old PRs). It's not that we don't want to see the speed up here, but we're skittish about making changes to the linter, especially multiple changes at once (in here we've got style changes, a structural change, and what looks to be some jsdoc changes—it's not clear they're all strictly dependent on each other).
I don't think we can merge this as is, so I propose closing this and approaching these things from a different angle.

Looking forward, we've got three suggestions:

- If you've got a change to the existing scripts, linters, etc., then they should be as small as possible. More PRs that do as little as possible is preferable. Counter-intuitive, but true.
- If an existing linting needs to change, then it should be accompanied by some tests (with refactoring as needed to make something testable—see https://github.com/mdn/browser-compat-data/pull/12673 for an example).
- If there's new linting to be done, then it should be implemented as a standalone test (for now, as in [`spec-urls.test.js`](https://github.com/mdn/browser-compat-data/blob/main/test/spec-urls.test.js)) or as part of the new linting architecture (an issue for which I still owe you, I think).

Let me know if there's any questions about this. Thank you!

**Comment by queengooborg** [comment]
I just saw this PR was closed.  I can understand the desire to keep changesets as small as possible, but to be honest, I don't see much room for stripping changes out.  There are a couple of things I could split into other PRs like the JSDoc and formatting changes to the main entry point, but the rest of the PR contains changes that all go together.

The problem I'm trying to solve with this PR is that right now, we have repeated lines of code for the same operation, causing us to iterate through the data multiple times.  What I'm attempting to do here is instead simplify this by introducing one function to iterate through the data and determine the parameters needed, which is then passed to another function that performs the check and logs any discrepancies.  While adding these new functions and replacing the old, I'm also describing them in JSDoc to give a more complete PR so we don't have to bother following up later.

Basically, all of the changes between lines 9 and 101, along with lines 110 through 125 (all on the "before" side) are all related and can't be cherry-picked, at least not in a sensible way.  The changes on other lines were included because they're so small on their own, they don't really warrant a separate PR, and we're already here working on this file.

File: test/linter/test-descriptions.js [hunk 1]
```diff
@@ -6,123 +6,108 @@ const { Logger } = require('./utils.js');
...
-    if (methodName.endsWith('_event')) {
-      const event = apiData[methodName];
-      const eventName = methodName.replace('_event', '');
-      if (event.__compat.description !== `<code>${eventName}</code> event`) {
-        logger.error(chalk`{red Incorrect event description for {bold ${apiName}#${methodName}}}
-      {yellow Actual: {bold "${event.__compat.description || ''}"}}
-      {green Expected: {bold "<code>${eventName}</code> event"}}`);
-      }
+function processApiData(apiData, apiName, logger) {
+  for (let methodName in apiData) {
```
Comment (caugner) [inline test/linter/test-descriptions.js:36]: This can be `const`:

```suggestion
  for (const methodName in apiData) {
```

File: test/linter/test-descriptions.js [hunk 2]
```diff
@@ -6,123 +6,108 @@ const { Logger } = require('./utils.js');
...
+    const method = apiData[methodName];
```
Comment (caugner) [inline test/linter/test-descriptions.js:37]: Isn't this `feature` and `featureName` rather than `method` and `methodName`?

File: test/linter/test-descriptions.js [hunk 3]
```diff
@@ -6,123 +6,108 @@ const { Logger } = require('./utils.js');
...
-function hasValidConstrutorDescription(apiData, apiName, logger) {
-  const constructor = apiData[apiName];
-  if (
-    constructor &&
-    constructor.__compat.description !== `<code>${apiName}()</code> constructor`
-  ) {
-    logger.error(chalk`{red Incorrect constructor description for {bold ${apiName}()}}
-      {yellow Actual: {bold "${constructor.__compat.description || ''}"}}
-      {green Expected: {bold "<code>${apiName}()</code> constructor"}}`);
+function checkError(error_type, name, method, expected, logger) {
```
Comment (caugner) [inline test/linter/test-descriptions.js:18]: 1. Would suggest `check[Method/Feature]Description` instead of `checkError`, and renaming `error_type` to `rule_name` or similar.
2. In general, would it make sense to wrap all these functions in a `class` and access the logger with `this.logger`?
3. The function parameter casing is inconsistent in this PR, compare `processApiData(apiData, apiName, logger)` and `checkError(error_type, name, method, expected, logger)`.

**Review by caugner — COMMENTED** [review]
LGTM, just some nits.

File: test/linter/test-descriptions.js (hunk 3 above)
Comment (queengooborg) [reply test/linter/test-descriptions.js]: A class would be nice, however I'd like to hold off on doing any further refactoring until the next-gen linter, just to save extra effort in the meantime!

Comment (caugner) [reply test/linter/test-descriptions.js]: Good point. 👍

File: test/linter/test-descriptions.js [hunk 4]
```diff
@@ -6,123 +6,108 @@ const { Logger } = require('./utils.js');
...
 
 /**
- * @param {Identifier} apiData
- * @param {String} apiName
- * @param {Logger} logger
+ * Check for errors in the description of a specified statement's description and return whether there's an error and log as such
+ *
+ * @param {string} ruleName The name of the error
+ * @param {string} name The name of the API method
+ * @param {Identifier} method The method's compat data
```
Comment (caugner) [inline test/linter/test-descriptions.js:13]: As established, `method` is actually more general a `feature` now (guess the check was originally used for methods only).

```suggestion
 * @param {Identifier} feature - The feature's compat data.
```

**Review by caugner — CHANGES_REQUESTED** [review]
One last follow-up nit, but still looking good, of course.
